shared/middleware/files_api.rb

```diff
@@ -400,6 +400,18 @@ def put_file(endpoint, encrypted_channel_id, filename, body)
       return bad_request if share_failure && share_failure[:type] != "address"
     end
 
+    # Don't allow project to be saved if it contains non-UTF-8 characters (causing error / project to not load when opened).
+    if 'sources' == endpoint
+      body_json = JSON.parse(body)
+      source = body_json["source"]
+      html = body_json["html"]
+
+      source_is_valid = source && source.force_encoding("UTF-8").valid_encoding?
+      html_is_valid = html ? source.force_encoding("UTF-8").valid_encoding? : true # HTML only exists for AppLab projects
+
+      return bad_request unless source_is_valid && html_is_valid
+    end
```

Comment:
I was looking for existing unit tests for `files_api` but couldn't find any. Definitely open to suggestions on if / how best to test this change.

Reply:
found them through some broken unit tests in the CI pipeline. Adding now haha